### ISSUE TITLE
Support for ghc-8.8.3

### ISSUE THREAD

**Issue by jneira** [haskell/haskell-ide-engine#1670]
* Finally we have a ghc working version for windows! :smile: 
* Package managers:
  * Already available for linux via ppa (@hvr is always the first one!)
  * It still is not in homebrew: https://github.com/Homebrew/homebrew-core/pull/50677
  * Available in chocolatey but it needs maintainer corrective action: https://chocolatey.org/packages/ghc/8.8.3
  * stackage nightly still is in ghc-8.8.2: https://www.stackage.org/nightly-2020-02-27
* I am building it locally in windows and i am preparing a pr to add the new version to azure: linux-cabal and windows-cabal

**Comment by jneira** [comment]
* Lot of failing tests in my windows local :disappointed: 
* Pr to correct chocolatey package: https://github.com/Mistuke/GhcChoco/pull/6

**Comment by jneira** [comment]
Link to failing tests in my local windows: https://gist.github.com/jneira/9bf777438bc0adbd5d6ba1e9be9350bf

**Comment by fendor** [comment]
It looks like all these bugs have a common root. Not finding the context implies some problem with the typechecked module, or parsed source or something. However, since the diagnostics work I suspect the latter.

**Comment by FirstLoveLife** [comment]
Hi, hie newbiee here. 

8.8.3 is already in homebrew. Hope to see hie support of ghc 8.8.3 soon. arch linux has also updated to 8.8.3 one week ago, maybe that's why hie is not work properly on my emacs.

**Comment by mouse07410** [comment]
I'm trying to build on MacOS Catalina 10.15.3 - the build fails:
```
$ ./cabal-hie-install 8.8.3
. . . . .
Configuring library for hie-install-0.8.1.0..
Preprocessing library for hie-install-0.8.1.0..
Building library for hie-install-0.8.1.0..
[2 of 8] Compiling Print            ( src/Print.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Print.o ) [Development.Shake changed]
[4 of 8] Compiling Stack            ( src/Stack.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Stack.o ) [Development.Shake.FilePath changed]
[5 of 8] Compiling Env              ( src/Env.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Env.o ) [Control.Monad.Extra changed]
[6 of 8] Compiling Help             ( src/Help.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Help.o ) [Development.Shake changed]
[7 of 8] Compiling Cabal            ( src/Cabal.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Cabal.o ) [Cabal.Config changed]
[8 of 8] Compiling HieInstall       ( src/HieInstall.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/HieInstall.o ) [Development.Shake changed]
Configuring executable 'script' for fake-package-0..
Preprocessing executable 'script' for fake-package-0..
Building executable 'script' for fake-package-0..
[1 of 1] Compiling Main             ( Main.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/fake-package-0/x/script/build/script/script-tmp/Main.dyn_o )
Linking /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/fake-package-0/x/script/build/script/script ...
Error when running Shake build system:
  at want, called at src/Development/Shake/Internal/Args.hs:83:69 in shk-0.18.5-7343d784:Development.Shake.Internal.Args
* Depends on: 8.8.3
  at error, called at src/Development/Shake/Internal/Rules/File.hs:179:58 in shk-0.18.5-7343d784:Development.Shake.Internal.Rules.File
* Raised the exception:
Error, file does not exist and no rule available:
  8.8.3

```

Any help?

**Comment by jneira** [comment]
@mouse07410 the command should be `cabal-hie-install hie-8.8.3`, you can check available targets with `cabal-hie-install help`. Let us know if the help info is not clear enough.

**Comment by mouse07410** [comment]
```
$ ./cabal-hie-install help
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
In order, the following will be built (use -v for more details):
 - hie-install-0.8.1.0 (lib) (configuration changed)
 - fake-package-0 (exe:script) (configuration changed)
Configuring library for hie-install-0.8.1.0..
Preprocessing library for hie-install-0.8.1.0..
Building library for hie-install-0.8.1.0..
[2 of 8] Compiling Print            ( src/Print.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Print.o ) [Development.Shake changed]
[4 of 8] Compiling Stack            ( src/Stack.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Stack.o ) [Development.Shake.FilePath changed]
[5 of 8] Compiling Env              ( src/Env.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Env.o ) [Control.Monad.Extra changed]
[6 of 8] Compiling Help             ( src/Help.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Help.o ) [Development.Shake changed]
[7 of 8] Compiling Cabal            ( src/Cabal.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/Cabal.o ) [Cabal.Config changed]
[8 of 8] Compiling HieInstall       ( src/HieInstall.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/build/HieInstall.o ) [Development.Shake changed]
Preprocessing library for hie-install-0.8.1.0..
Running Haddock on library for hie-install-0.8.1.0..
Haddock coverage:
Warning: 'stack' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'ghc' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'cabal' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
  25% (  1 /  4) in 'HieInstall'
  Missing documentation for:
    Module header
    defaultMain (src/HieInstall.hs:14)
    buildIcuMacosFix (src/HieInstall.hs:98)
Warning: HieInstall: could not find link destinations for:
    VersionNumber
Documentation created:
/Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/hie-install-0.8.1.0/doc/html/hie-install/index.html
Configuring executable 'script' for fake-package-0..
Preprocessing executable 'script' for fake-package-0..
Building executable 'script' for fake-package-0..
[1 of 1] Compiling Main             ( Main.hs, /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/fake-package-0/x/script/build/script/script-tmp/Main.dyn_o )
Linking /Users/ur20980/src/haskell-ide-engine/dist-newstyle/build/x86_64-osx/ghc-8.8.3/fake-package-0/x/script/build/script/script ...

Usage:
    stack install.hs <target> [options]
    or
    cabal v2-run install.hs --project-file install/shake.project -- <target> [options]

Targets:
    help                Show help message including all targets
                        
    hie                 Install hie with the latest available GHC and the data files
    latest              Install hie with the latest available GHC
    data                Get the required data-files for `hie` (Hoogle DB)
    hie-8.6.5           Install hie for GHC version 8.6.5
    hie-8.8.2           Install hie for GHC version 8.8.2
    hie-8.8.3           Install hie for GHC version 8.8.3
                        
    ghcs                Show all GHC versions that can be installed via `cabal-build`.
                        
    icu-macos-fix       Fixes icu related problems in MacOS

Options:
    -j[N], --jobs[=N]   Allow N jobs/threads at once [default number of CPUs].
    -s, --silent        Don't print anything.
    -q, --quiet         Print less (pass repeatedly for even less).
    -V, --verbose       Print more (pass repeatedly for even more).

Build completed in 0.01s

$ 
```
*will update this with the results when the build (`./cabal-hie-install hie-8.8.3`) completes.*

**Comment by mouse07410** [comment]
Also, I'm surprised that even when explicitly specifying building for `ghc-8.8.3` *and* having `ghc-8.8.3` the system default, `stack` still seems to have built `hie` with `ghc-8.8.2`:

```sh
$ stack ./install.hs hie-8.8.3
# stack (for check)
# git (for submodules)
# git (for submodules)
Options:
    Verbosity level: Info
# stack (for hie-8.8.3)
Stack has not been tested with GHC versions above 8.6, and using 8.8.2, this may fail
Stack has not been tested with Cabal versions above 2.4, but version 3.0.1.0 was found, this may fail
Copying from /Users/ur20980/src/haskell-ide-engine/.stack-work/install/x86_64-osx/22ab76dd0602426768f3f9456fbe043d018a7a8f4d8bb38928967d7287145f92/8.8.2/bin/hie to /Users/ur20980/.local/bin/hie
Copying from /Users/ur20980/src/haskell-ide-engine/.stack-work/install/x86_64-osx/22ab76dd0602426768f3f9456fbe043d018a7a8f4d8bb38928967d7287145f92/8.8.2/bin/hie-wrapper to /Users/ur20980/.local/bin/hie-wrapper

Copied executables to /Users/ur20980/.local/bin:
- hie
- hie-wrapper
# stack (for hie-8.8.3)
Stack has not been tested with GHC versions above 8.6, and using 8.8.2, this may fail
Stack has not been tested with Cabal versions above 2.4, but version 3.0.1.0 was found, this may fail
Stack has not been tested with GHC versions above 8.6, and using 8.8.2, this may fail
Stack has not been tested with Cabal versions above 2.4, but version 3.0.1.0 was found, this may fail
Build completed in 1.21s

$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.8.3
$ ghcup list
Available versions:
✗ ghc	7.10.3	7.10
✗ ghc	8.0.2	8.0
✗ ghc	8.2.2	8.2
✗ ghc	8.4.1
✗ ghc	8.4.2
✗ ghc	8.4.3
✗ ghc	8.4.4	8.4
✗ ghc	8.6.1	bad
✗ ghc	8.6.2
✗ ghc	8.6.3
✗ ghc	8.6.4
✔ ghc	8.6.5	recommended,8.6
✗ ghc	8.8.1
✔ ghc	8.8.2
✔ ghc	8.8.3	latest,8.8 [in blue]
$ 
```

**Comment by jneira** [comment]
> stack still seems to have built hie with ghc-8.8.2

I think this is a bug in the stack output, you can check the real version with `hie --version`

As hie is not fully supporting `ghc-8.3.3` we still have not added a `cabal.project-8.8.3` and that is the reason for the failed build with `cabal-hie-install hie-8.8.3`. Try `cp cabal.project-8.8.2 cabal.project-8.8.3` and the build may be successful.

**Comment by mouse07410** [comment]
>> stack still seems to have built hie with `ghc-8.8.2`
>
> I think this is a bug in the stack output, you can check the real version with `hie --version`

I'm afraid `stack` output is correct:
```sh
$ type hie
hie is /Users/ur20980/.local/bin/hie
$ type hie-8.8.3
hie-8.8.3 is /Users/ur20980/.local/bin/hie-8.8.3
$ stack --version
Version 2.1.3, Git revision 0fa51b9925decd937e4a993ad90cb686f88fa282 (7739 commits) x86_64 hpack-0.31.2
$ hie --version
Version 1.2, Git revision ee5b98c6bd3610ff90cb5373608a09c5813dcb69 (dirty) (3829 commits) x86_64 ghc-8.8.2
$ hie-8.8.3 --version
Version 1.2, Git revision ee5b98c6bd3610ff90cb5373608a09c5813dcb69 (dirty) (3829 commits) x86_64 ghc-8.8.2
$ type ghc
ghc is /Users/ur20980/.ghcup/bin/ghc
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.8.3
$ 
```

Also, from the following output in my small sample project directory it seems clear that `stack` uses 8.8.2, no matter what:
```sh
$ ls -R .stack-work | fgrep 8.8.3
$ ls -R .stack-work | fgrep 8.8.2
libHSStr2Split-0.1.0.0-2eOuLbeNvAA1cmgGTGD6DW-ghc8.8.2.dylib
8.8.2
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2:
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/bin:
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/doc:
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/doc/Str2Split-0.1.0.0:
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/lib:
x86_64-osx-ghc-8.8.2
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/lib/x86_64-osx-ghc-8.8.2:
libHSStr2Split-0.1.0.0-2eOuLbeNvAA1cmgGTGD6DW-ghc8.8.2.dylib
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/lib/x86_64-osx-ghc-8.8.2/Str2Split-0.1.0.0-2eOuLbeNvAA1cmgGTGD6DW:
.stack-work/install/x86_64-osx/a9cf3cc2cddd3cc7bddda2de466f06e3510275e1f229024b61ba5e90415b5b40/8.8.2/pkgdb:
```

It's not a bug in the output - it's the compiler that `stack` picked. I'm trying to understand why (adding screenshot, as only the color tells which of the installed compilers is set as default):

<img width="292" alt="Screen Shot 2020-03-09 at 21 54 36" src="https://user-images.githubusercontent.com/5923577/76272030-94b4bd00-6250-11ea-8e85-e37fecf8f2de.png">

**Comment by mouse07410** [comment]
> Try `cp cabal.project-8.8.2 cabal.project-8.8.3` and the build may be successful.

Unfortunately, it isn't:
```
$ ./cabal-hie-install hie-8.8.3
. . . . .
Completed    haskell-lsp-types-0.20.0.0 (lib)
Starting     haskell-lsp-0.20.0.1 (lib)
Building     haskell-lsp-0.20.0.1 (lib)
Haddock      haskell-lsp-0.20.0.1 (lib)
Installing   haskell-lsp-0.20.0.1 (lib)
Completed    haskell-lsp-0.20.0.1 (lib)
Starting     hie-plugin-api-1.2 (lib)
Building     hie-plugin-api-1.2 (lib)
Haddock      hie-plugin-api-1.2 (lib)
Warning: Failed to build documentation for hie-plugin-api-1.2 (which is
Symlinking 'hie-wrapper'
required by exe:hie from haskell-ide-engine-1.2 and exe:hie-wrapper from
haskell-ide-engine-1.2).
Symlinking 'hie'
Error when running Shake build system:
  at want, called at src/Development/Shake/Internal/Args.hs:83:69 in shk-0.18.5-78bc9188:Development.Shake.Internal.Args
* Depends on: hie-8.8.3
* Raised the exception:
/Users/ur20980/.cabal/bin/hie: copyFile:atomicCopyFileContents:withReplacementFile:copyFileToHandle:openBinaryFile: does not exist (No such file or directory)

```

**Comment by jneira** [comment]
Weird indeed, could you double check that the `hie-8.8.3` executed is the one from cabal and no from stack? Executables from cabal usually does not have the git revision:
```
PS D:\dev\ws\haskell\ghcide> hie-8.8.3 --version
Version 1.1 x86_64 ghc-8.8.3
```

**Comment by mouse07410** [comment]
> Weird indeed, could you double check that the `hie-8.8.3` executed is the one from cabal and no from stack?

Darn, you're right!
```sh
$ ~/.cabal/bin/hie-8.8.3 --version
Version 1.2 x86_64 ghc-8.8.3
```

> Is it possible that you created a `stack-8.8.3.yaml` (needed to install hie with that version with `stack install.hs hie-8.8.3`) from the default `stack.yaml` or `stack-8.8.2.yaml` but you did not change the resolver?

Well, sort of... I copied `stack-8.8.2.yaml` to `stack-8.8.3.yaml` and changed the resolver to `lts-15.3`:
```sh
$ hie-8.8.3 --version
Version 1.2, Git revision ee5b98c6bd3610ff90cb5373608a09c5813dcb69 (dirty) (3829 commits) x86_64 ghc-8.8.2
$ ll stack-8.8.3.yaml
-rw-r--r--  1 ur20980  MITLL\Domain Users  1061 Mar  9 22:00 stack-8.8.3.yaml
$ head stack-8.8.3.yaml
resolver: lts-15.3
packages:
- .
- hie-plugin-api

extra-deps:
# - ./submodules/HaRe

- apply-refact-0.7.0.0
- brittany-0.12.1.1
```
What should the resolver be set to for this build in `stack-8.8.3.yaml`?

> And for stack, it does not use the `ghc` on path by default but it downloads and uses a sandboxed one.

But... my `~/.stack/config` defines
```yaml
$ head -20 ~/.stack/config.yaml 
. . . . .

system-ghc: true
skip-ghc-check: true
. . . . .
```

Do you have any idea why `stack` also appears to build my own sample project with GHC-8.8.2 instead of GHC-8.8.3?

**Comment by jneira** [comment]
Well, `lts-15.3` is for `ghc-8.8.2` so using that resolver makes stack ignore the system `ghc-8.8.3` (cause versions dont match).
There is no lts nor nightly resolver for `ghc-8.8.3` right now so the unique way i can think of is set
```
resolver: ghc-8.8.3
system-ghc: true
skip-ghc-check: true
```
(`system-ghc and `skip-ghc-check` could be in the global config)

**Comment by jneira** [comment]
@FirstLoveLife well maybe it can be another root cause, have you tried another ghc version that works with hie to check it it works?

**Comment by mouse07410** [comment]
> Well, lts-15.3 is for `ghc-8.8.2` so using that resolver makes stack ignore the system `ghc-8.8.3` (cause versions dont match). There is no lts nor nightly resolver for `ghc-8.8.3`

Interesting. But `stack` *did* build `hie-8.8.3`, though it used `ghc-8.8.2` for that. Does it mean that VS Code can use `hie-8.8.3` for projects that require GHC-8.8.3? 

I.e., should I worry about which GHC compiler stack used to build `hie-8.8.3`?

**Comment by jneira** [comment]
> But stack did build hie-8.8.3, though it used ghc-8.8.2 for that. Does it mean that VS Code can use hie-8.8.3 for projects that require GHC-8.8.3?

I am afraid it does not. The hie version comes from yaml files using a convention: it assumes the version in the `stack-${version}.yaml` file name matches the ghc version of its resolver. So you could create a `stack-666.yaml` and the install script would create a `hie-666` executable but the ghc will continue being the defined in the resolver.

> should I worry about which GHC compiler stack used to build hie-8.8.3?

Yeah. The existent `stack.yaml`'s fullfill the convention but you have to ensure it manually if you creates a new `stack.yaml`

**Comment by gdziadkiewicz** [comment]
Nightly with ghc 8.8.3 appeared - https://www.stackage.org/nightly-2020-03-10

**Comment by FirstLoveLife** [comment]
> @FirstLoveLife well maybe it might be another root cause, have you tried another ghc version that works with hie to check it it works?

Not yet. I only have manjaro now. It is not really easy to downgrade to 8.8.2 given the dependencies

**Comment by dschrempf** [comment]
FYI, Stackage resolvers lts-15.4 and lts-15.5 support ghc 8.8.3 now!

**Comment by mouse07410** [comment]
Yes that is good to know.

However my current problem is building with and for ghc-8.10.1.

@jneira any comments? Cabal build would be just fine too.

**Comment by jneira** [comment]
Well, [the branch i started to add full support to ghc-8.8.3](https://github.com/haskell/haskell-ide-engine/pull/1697) can be built with cabal and stack. The test suite has a bunch of errors but i think it could be usable.
I'll open a new issue to track ghc-8.10 separately.

**Comment by dschrempf** [comment]
@jneira I just tried buiding hie with your branch `ghc-8.8.3`, and it works fine! It also works in the editor apparently, but I did not use it extensively yet. Thanks!

**Comment by mouse07410** [comment]
> ...just tried building hie with your branch `ghc-8.8.3`, and it works fine...

@dschrempf could you clarify? There is no such branch in the HIE repo. Did you mean you "built **`master`** branch with GHC-8.8.3 and it worked"?

FWIW, I built master with GHC-8.8.3 and it seems to be working, under VS Code and HLS plugin.

@jneira to declare full GHC-8.8.3 support, you're waiting for #1697 to pass CI tests and merge?

**Comment by dschrempf** [comment]
I was referring to the fork of @jneira, which has a branch https://github.com/jneira/haskell-ide-engine/tree/ghc-8.8.3.

**Comment by jneira** [comment]
@mouse07410 yeah, i cant promise it but i'll try to take a look to failing tests

**Comment by mouse07410** [comment]
> I cant promise it but I'll try to take a look to failing tests

Understood, thanks!

**Comment by tofutim** [comment]
when i try to ./cabal-hie-install hie-8.8.3 on macos, I get
```
cabal: Could not resolve dependencies:
[__0] trying: haskell-ide-engine-1.4 (user goal)
[__1] next goal: hie-bios (dependency of haskell-ide-engine)
[__1] rejecting: hie-bios-0.4.0 (conflict: haskell-ide-engine => hie-bios>=0.5
&& <0.6.0)
[__1] skipping: hie-bios-0.3.2, hie-bios-0.3.1, hie-bios-0.3.0,
hie-bios-0.2.1, hie-bios-0.2.0, hie-bios-0.1.1, hie-bios-0.1.0 (has the same
characteristics that caused the previous version to fail: excluded by
constraint '>=0.5 && <0.6.0' from 'haskell-ide-engine')
[__1] fail (backjumping, conflict set: haskell-ide-engine, hie-bios)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: haskell-ide-engine, hie-bios
```

**Comment by fendor** [comment]
I think you have to do a `cabal update`? Maybe your package index isnt up-to-date.
